### ISSUE TITLE
Sealevel contracts: instruction discriminators should be used when appropriate

### ISSUE THREAD

**Issue by tkporter** [hyperlane-xyz/hyperlane-monorepo#2253]
The typical pattern for sealevel contract instructions is to just serialize/deserialize an `enum Instruction` using Borsh. Borsh enum (de)serialization works by just looking at what index the enum variant is, and encoding that into the first byte. With this type of (de)serialization by itself, it'd mean that we'd need to require recipient contracts to have specific instruction variants at particular indices in their instruction enums. This isn't really tenable

What we should instead do, is something closer to function signatures in the EVM world. In Solana, they're called discriminants. See https://github.com/solana-labs/solana-program-library/blob/master/token/transfer-hook-interface/src/instruction.rs#L49 as a good example of this

Basically the first few bytes of the instruction data should be a hash that's specific to the instruction type. We'd want a unique one for recipient.ISM, ISM.verify, ISM.type, recipient.handle, any ISM-specific things like `validators_and_threshold`, etc

**Comment by tkporter** [comment]
Done
